transformer_engine/pytorch/distributed.py

```diff
@@ -137,25 +141,32 @@ def gather_split_1d_tensor(
     return gathered
 
 
-@contextmanager
-def activation_recompute_forward(
-    activation_recompute: bool = False,
-    recompute_phase: bool = False,
-) -> None:
+class activation_recompute_forward(AbstractContextManager, ContextDecorator):
```

Comment:
Curious, why?

Reply:
The native PyTorch re-entrant checkpoint has an option to provide a context function that returns `(forward_ctx, recompute_ctx)` objects, which are then combined with the native `_checkpoint_hook` and `_recomputation_hook` within PyTorch. Re-implementing our `activation_recompute_forward()` as a class instead of a function made this workflow easier/cleaner for me.

Of course this PR does not use the native PyTorch checkpoint, but I've confirmed in my limited testing that it does work when supplied with the right context function. The caveat is that you have to set the RNG states once in the beginning correctly for all the devices, and then make sure the modules never tamper with the RNG state themselves. The `preserve_rng_state=True` option in PyTorch's native checkpointing makes sure that the initially correct RNG states are preserved through the checkpoint and recompute.

I kept this out of the PR because:
1. It does not work with TE modules/models that use `CudaRNGStateManager`. Supporting that requires TE to implement its own non-reentrant checkpoint, which I've done in this PR.
2. TE modules that implement custom forward/backward ops do not benefit from the early stopping feature in the PyTorch checkpoint because they call `ctx.save_for_backward()` only once, with all the saved tensors passed in bulk. The internal bookkeeping for early stopping relies on this being called separately for each tensor that needs to be saved.

Since I already did this conversion, I left it in this PR in preparation for the future possibility that we might figure out a way to eliminate `CudaRNGStateManager` and perhaps restructure how we use `ctx.save_for_backward()` in custom forward/backward ops. That would let us completely get rid of TE's own checkpoint implementation and just use PyTorch's native API.